File: mycousinvinyl/backend/app/entrypoints/http/schemas/collection.py
"""
Collection API schemas.

SECURITY NOTE: user_id is never included in request schemas - it's always
extracted from the authenticated user context.
"""

from typing import Optional, List
from uuid import UUID
from datetime import datetime, date
from decimal import Decimal
from pydantic import BaseModel, Field, condecimal

from app.domain.value_objects import Condition


class CollectionItemBase(BaseModel):
    """Base collection item schema (for requests)."""
    model_config = {'populate_by_name': True, 'from_attributes': True}

    pressing_id: UUID = Field(..., description="Pressing ID to add to collection")
    media_condition: Condition = Field(..., example=Condition.NEAR_MINT, description="Vinyl condition")
    sleeve_condition: Condition = Field(..., example=Condition.VG_PLUS, description="Sleeve condition")
    purchase_price: Optional[condecimal(ge=Decimal('0'), decimal_places=2)] = Field(
        None, example=Decimal('25.99'), description="Purchase price"
    )
    purchase_currency: Optional[str] = Field(
        None, min_length=3, max_length=3, example="DKK", description="ISO 4217 currency code"
    )
    purchase_date: Optional[date] = Field(None, example="2024-01-15")
    seller: Optional[str] = Field(None, max_length=200, example="Discogs seller")
    location: Optional[str] = Field(
        None,
        max_length=200,
        example="Shelf A3",
        alias="storage_location",
        serialization_alias="location",
    )
    defect_notes: Optional[str] = Field(None, example="Minor scuff on side B")
    notes: Optional[str] = Field(
        None,
        example="First pressing with poster",
        alias="user_notes",
        serialization_alias="notes",
    )


class CollectionItemCreate(CollectionItemBase):
    """Schema for adding item to collection."""
    pass


class CollectionItemUpdate(BaseModel):
    """Schema for updating collection item (all fields optional)."""
    model_config = {'populate_by_name': True}

    media_condition: Optional[Condition] = None
    sleeve_condition: Optional[Condition] = None
    purchase_price: Optional[condecimal(ge=Decimal('0'), decimal_places=2)] = None
    purchase_currency: Optional[str] = Field(None, min_length=3, max_length=3)
    purchase_date: Optional[date] = None
    seller: Optional[str] = Field(None, max_length=200)
    location: Optional[str] = Field(
        None,
        max_length=200,
        alias="storage_location",
        serialization_alias="location",
    )
    defect_notes: Optional[str] = None
    notes: Optional[str] = Field(None, alias="user_notes", serialization_alias="notes")


class ConditionUpdateRequest(BaseModel):
    """Schema for updating condition-specific fields."""
    media_condition: Optional[Condition] = None
    sleeve_condition: Optional[Condition] = None
    defect_notes: Optional[str] = None


class PurchaseInfoUpdateRequest(BaseModel):
    """Schema for updating purchase information."""
    purchase_price: Optional[condecimal(ge=Decimal('0'), decimal_places=2)] = None
    purchase_currency: Optional[str] = Field(None, min_length=3, max_length=3)
    purchase_date: Optional[date] = None
    seller: Optional[str] = Field(None, max_length=200)


class RatingUpdateRequest(BaseModel):
    """Schema for updating rating and notes."""
    rating: int = Field(..., ge=0, le=5, example=5, description="Rating from 0-5")
    notes: Optional[str] = Field(None, example="Classic album, sounds amazing")


class CollectionItemResponse(CollectionItemBase):
    """Schema for collection item responses."""
    id: UUID
    user_id: UUID
    rating: Optional[int] = Field(None, ge=0, le=5)
    play_count: int = Field(default=0, ge=0)
    last_played: Optional[datetime] = None
    date_added: datetime
    updated_at: datetime


class ArtistSummary(BaseModel):
    """Lightweight artist summary for collection views."""
    id: UUID
    name: str
    sort_name: Optional[str] = None
    country: Optional[str] = None

    class Config:
        from_attributes = True


class AlbumSummary(BaseModel):
    """Lightweight album summary for collection views."""
    id: UUID
    title: str
    release_year: Optional[int] = None
    genres: List[str] = Field(default_factory=list)
    image_url: Optional[str] = None

    class Config:
        from_attributes = True


class MarketDataSummary(BaseModel):
    """Market pricing data summary."""
    min_value: Optional[Decimal] = None
    median_value: Optional[Decimal] = None
    max_value: Optional[Decimal] = None
    currency: Optional[str] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class CollectionItemDetailResponse(BaseModel):
    """Enhanced collection item response with artist and album details."""
    model_config = {'populate_by_name': True, 'from_attributes': True}

    id: UUID
    user_id: UUID
    pressing_id: UUID
    pressing_image_url: Optional[str] = None
    media_condition: Condition
    sleeve_condition: Condition
    purchase_price: Optional[Decimal] = None
    purchase_currency: Optional[str] = None
    purchase_date: Optional[date] = None
    seller: Optional[str] = None
    location: Optional[str] = Field(None, alias="storage_location", serialization_alias="location")
    defect_notes: Optional[str] = None
    notes: Optional[str] = Field(None, alias="user_notes", serialization_alias="notes")
    rating: Optional[int] = Field(None, ge=0, le=5)
    play_count: int = Field(default=0, ge=0)
    last_played: Optional[datetime] = None
    date_added: datetime
    updated_at: datetime
    # Enriched data
    artist: ArtistSummary
    album: AlbumSummary
    market_data: Optional[MarketDataSummary] = None


class CollectionItemSummary(BaseModel):
    """Lightweight collection item summary."""
    model_config = {'from_attributes': True}

    id: UUID
    pressing_id: UUID
    media_condition: Condition
    sleeve_condition: Condition
    date_added: datetime


class CollectionSearchParams(BaseModel):
    """Query parameters for collection search and filtering."""
    query: Optional[str] = Field(None, min_length=1, description="Search album titles and artists")
    media_conditions: Optional[List[Condition]] = Field(None, description="Filter by media conditions (OR logic)")
    sleeve_conditions: Optional[List[Condition]] = Field(None, description="Filter by sleeve conditions (OR logic)")
    rating_min: Optional[int] = Field(None, ge=0, le=5, description="Minimum rating")
    rating_max: Optional[int] = Field(None, ge=0, le=5, description="Maximum rating")
    price_min: Optional[Decimal] = Field(None, ge=0, description="Minimum purchase price")
    price_max: Optional[Decimal] = Field(None, ge=0, description="Maximum purchase price")
    sort_by: str = Field(
        default="date_added_desc",
        description="Sort field: date_added_desc, date_added_asc, artist, album, rating_desc, price_desc"
    )
    limit: int = Field(default=100, ge=1, le=1000)
    offset: int = Field(default=0, ge=0)


class TopArtistEntry(BaseModel):
    """Top artist entry for collection statistics."""
    artist_id: UUID
    artist_name: str
    collected_count: int


class TopAlbumEntry(BaseModel):
    """Top album entry for collection statistics."""
    album_id: UUID
    album_title: str
    artist_id: UUID
    artist_name: str
    collected_count: int


class CollectionStatistics(BaseModel):
    """Collection statistics response."""
    total_albums: int = Field(..., example=150)
    total_purchase_price: float = Field(..., example=3500.00)
    min_value: float = Field(..., example=5.00)
    avg_value: float = Field(..., example=23.33)
    max_value: float = Field(..., example=150.00)
    low_est_sales_price: float = Field(..., example=2800.00)
    avg_est_sales_price: float = Field(..., example=4200.00)
    high_est_sales_price: float = Field(..., example=6500.00)
    currency: str = Field(..., example="DKK")
    top_artists: List[TopArtistEntry] = Field(default_factory=list)
    top_albums: List[TopAlbumEntry] = Field(default_factory=list)

    class Config:
        from_attributes = True


class AlbumPlayIncrementResponse(BaseModel):
    """Response for incrementing album play count."""
    album_id: UUID
    play_count: int
    play_count_ytd: int
    last_played_at: datetime


class PlayedAlbumEntry(BaseModel):
    """Played album entry for YTD stats."""
    album_id: UUID
    album_title: str
    artist_id: UUID
    artist_name: str
    play_count_ytd: int
    last_played_at: Optional[datetime] = None

File: mycousinvinyl/frontend/src/api/services/collection.ts
/**
 * Collection API service.
 */

import apiClient from '../client';
import type {
  CollectionItemResponse,
  CollectionItemCreate,
  CollectionItemUpdate,
  CollectionStatistics,
  PaginatedResponse,
  MessageResponse,
  Condition,
  CollectionItemDetailResponse,
  CollectionImportResponse,
  AlbumPlayIncrementResponse,
  PlayedAlbumEntry,
} from '@/types/api';

export const collectionApi = {
  /**
   * Add pressing to collection.
   */
  addItem: async (data: CollectionItemCreate): Promise<CollectionItemResponse> => {
    const response = await apiClient.post<CollectionItemResponse>('/api/v1/collection', data);
    return response.data;
  },

  /**
   * Get collection item by ID.
   */
  getById: async (id: string): Promise<CollectionItemResponse> => {
    const response = await apiClient.get<CollectionItemResponse>(`/api/v1/collection/${id}`);
    return response.data;
  },

  /**
   * Get user's collection with filters and search.
   */
  getCollection: async (params?: {
    query?: string;
    media_conditions?: Condition[];
    sleeve_conditions?: Condition[];
    rating_min?: number;
    rating_max?: number;
    sort_by?: string;
    limit?: number;
    offset?: number;
  }): Promise<PaginatedResponse<CollectionItemResponse>> => {
    const response = await apiClient.get<PaginatedResponse<CollectionItemResponse>>(
      '/api/v1/collection',
      { params }
    );
    return response.data;
  },

  /**
   * Get user's collection with artist and album details.
   * Returns data pre-sorted for hierarchical display.
   */
  getCollectionWithDetails: async (params?: {
    query?: string;
    sort_by?: string;
    limit?: number;
    offset?: number;
  }): Promise<PaginatedResponse<CollectionItemDetailResponse>> => {
    const response = await apiClient.get<PaginatedResponse<CollectionItemDetailResponse>>(
      '/api/v1/collection/with-details',
      { params }
    );
    return response.data;
  },

  /**
   * Update collection item.
   */
  update: async (id: string, data: CollectionItemUpdate): Promise<CollectionItemResponse> => {
    const response = await apiClient.put<CollectionItemResponse>(`/api/v1/collection/${id}`, data);
    return response.data;
  },

  /**
   * Update item condition.
   */
  updateCondition: async (
    id: string,
    data: {
      media_condition?: Condition;
      sleeve_condition?: Condition;
      defect_notes?: string;
    }
  ): Promise<CollectionItemResponse> => {
    const response = await apiClient.put<CollectionItemResponse>(
      `/api/v1/collection/${id}/condition`,
      data
    );
    return response.data;
  },

  /**
   * Update purchase information.
   */
  updatePurchaseInfo: async (
    id: string,
    data: {
      purchase_price?: number;
      purchase_currency?: string;
      purchase_date?: string;
      seller?: string;
    }
  ): Promise<CollectionItemResponse> => {
    const response = await apiClient.put<CollectionItemResponse>(
      `/api/v1/collection/${id}/purchase`,
      data
    );
    return response.data;
  },

  /**
   * Update item rating.
   */
  updateRating: async (
    id: string,
    data: {
      rating: number;
      notes?: string;
    }
  ): Promise<CollectionItemResponse> => {
    const response = await apiClient.put<CollectionItemResponse>(
      `/api/v1/collection/${id}/rating`,
      data
    );
    return response.data;
  },

  /**
   * Increment play count.
   */
  incrementPlayCount: async (id: string): Promise<CollectionItemResponse> => {
    const response = await apiClient.post<CollectionItemResponse>(
      `/api/v1/collection/${id}/play`
    );
    return response.data;
  },

  /**
   * Remove item from collection.
   */
  removeItem: async (id: string): Promise<MessageResponse> => {
    const response = await apiClient.delete<MessageResponse>(`/api/v1/collection/${id}`);
    return response.data;
  },

  /**
   * Get collection statistics.
   */
  getStatistics: async (): Promise<CollectionStatistics> => {
    const response = await apiClient.get<CollectionStatistics>('/api/v1/collection/statistics');
    return response.data;
  },

  /**
   * Increment album play count.
   */
  incrementAlbumPlayCount: async (albumId: string): Promise<AlbumPlayIncrementResponse> => {
    const response = await apiClient.post<AlbumPlayIncrementResponse>(
      `/api/v1/collection/albums/${albumId}/play`
    );
    return response.data;
  },

  /**
   * Get played albums for the year.
   */
  getPlayedAlbumsYtd: async (params?: {
    year?: number;
    limit?: number;
    offset?: number;
  }): Promise<PaginatedResponse<PlayedAlbumEntry>> => {
    const response = await apiClient.get<PaginatedResponse<PlayedAlbumEntry>>(
      '/api/v1/collection/plays/ytd',
      { params }
    );
    return response.data;
  },

  /**
   * Import Discogs CSV collection export.
   */
  importDiscogs: async (file: File): Promise<CollectionImportResponse> => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await apiClient.post<CollectionImportResponse>(
      '/api/v1/collection/imports/discogs',
      formData,
      { headers: { 'Content-Type': 'multipart/form-data' } }
    );
    return response.data;
  },

  /**
   * Sync Discogs collection via API (PAT or OAuth).
   */
  syncDiscogs: async (): Promise<CollectionImportResponse> => {
    const response = await apiClient.post<CollectionImportResponse>(
      '/api/v1/collection/imports/discogs/sync'
    );
    return response.data;
  },

  /**
   * Get import status.
   */
  getImportStatus: async (importId: string): Promise<CollectionImportResponse> => {
    const response = await apiClient.get<CollectionImportResponse>(
      `/api/v1/collection/imports/${importId}`
    );
    return response.data;
  },
};
